File: server/datasets/java_source-code/Multiple-D1/structural_1.java
// QuickSort Suite: Structural Method Extraction & Reordering
// Author: Hannah (structural_1.java - Type 3 of Mary)
public class QuickSortSuite {
    private int[] data;
    private int comparisons;
    private int swaps;

    public QuickSortSuite(int[] input) {
        this.data = (input != null) ? input.clone() : new int[0];
        this.comparisons = 0;
        this.swaps = 0;
    }

    private static void performSwap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    public boolean isSorted() {
        int k = 0;
        while (k < this.data.length - 1) {
            if (this.data[k] > this.data[k + 1]) return false;
            k++;
        }
        return true;
    }

    public int[] sort(boolean inPlace) {
        int[] target = inPlace ? this.data : this.data.clone();
        if (target.length > 1) {
            quicksortRecursive(target, 0, target.length - 1);
        }
        return target;
    }

    private int partition(int[] arr, int low, int high) {
        int mid = (low + high) / 2;
        int pIdx = ((arr[low] <= arr[mid] && arr[mid] <= arr[high]) || (arr[high] <= arr[mid] && arr[mid] <= arr[low])) ? mid : low;
        performSwap(arr, pIdx, high);
        this.swaps++;
        
        int pivot = arr[high];
        int i = low - 1;
        for (int j = low; j < high; j++) {
            this.comparisons++;
            if (arr[j] <= pivot) {
                i++;
                performSwap(arr, i, j);
                this.swaps++;
            }
        }
        performSwap(arr, i + 1, high);
        this.swaps++;
        return i + 1;
    }

    private void quicksortRecursive(int[] arr, int low, int high) {
        if (low < high) {
            int pi = partition(arr, low, high);
            quicksortRecursive(arr, low, pi - 1);
            quicksortRecursive(arr, pi + 1, high);
        }
    }
}

File: server/datasets/python_source-code/Multiple-D1/structural_1.py
"""
QuickSort Suite: Method Extraction & Helper Restructuring
Author: Hannah (structural_1.py - Type 3 of Mary)
"""
from typing import List, Optional

def perform_swap(arr: List[int], i: int, j: int) -> None:
    temp = arr[i]
    arr[i] = arr[j]
    arr[j] = temp

def evaluate_bounds(a: int, b: int, c: int) -> bool:
    return (a <= b <= c) or (c <= b <= a)

class QuickSortSuite:
    def __init__(self, data: Optional[List[int]] = None):
        self.data = list(data) if data is not None else []
        self.comparisons = 0
        self.swaps = 0

    def _pick_pivot(self, arr: List[int], low: int, high: int) -> int:
        mid = (low + high) // 2
        if evaluate_bounds(arr[low], arr[mid], arr[high]):
            return mid
        if evaluate_bounds(arr[mid], arr[low], arr[high]):
            return low
        return high

    def _partition(self, arr: List[int], low: int, high: int) -> int:
        p_idx = self._pick_pivot(arr, low, high)
        perform_swap(arr, p_idx, high)
        self.swaps += 1
        pivot = arr[high]
        
        i = low - 1
        for j in range(low, high):
            self.comparisons += 1
            if arr[j] <= pivot:
                i += 1
                perform_swap(arr, i, j)
                self.swaps += 1
        
        perform_swap(arr, i + 1, high)
        self.swaps += 1
        return i + 1

    def _quicksort_recursive(self, arr: List[int], low: int, high: int) -> None:
        if low < high:
            pi = self._partition(arr, low, high)
            self._quicksort_recursive(arr, low, pi - 1)
            self._quicksort_recursive(arr, pi + 1, high)

    def sort(self, inplace: bool = True) -> List[int]:
        target = self.data if inplace else list(self.data)
        if len(target) > 1:
            self._quicksort_recursive(target, 0, len(target) - 1)
        return target

    def is_sorted(self) -> bool:
        return all(self.data[k] <= self.data[k + 1] for k in range(len(self.data) - 1))
